[IOperation unit testing] Add support for an OperationTreeVerifier - test …

mavasani opened the pull request:
…only OperationWalker to dump operation tree.

The dump approach should enable to us to (relatively) easily extend the existing C# and VB binding tests to verify the operation tree.
I have moved the existing IOperation tests to the dump approach and also added couple of additional tests to demonstrate how to write these tests for entire method body operation tree verification.

mavasani commented:
Tagging @JohnHamby @AlekseyTs @gafter @genlu @srivatsn for review.

mavasani commented:
Also tagging @ManishJayaswal

mavasani commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -40,7 +39,51 @@ static void Main()
+            CompileAndVerify(source: source, expectedOutput: "TwoZero")
+                .VerifyOperationTree(@"
+Test.Main()
+  IBlockStatement (8 statements, 6 locals) (OperationKind.BlockStatement)
```
Please provide comments on the dump, it probably seems verbose but I can't think of a smaller readable dump of the operation tree.

JohnHamby commented:
This is brilliant! This approach will greatly reduce the time needed to construct comprehensive tests.

srivatsn commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -40,7 +39,51 @@ static void Main()
-            CompileAndVerify(source: source, expectedOutput: "TwoZero");
+            CompileAndVerify(source: source, expectedOutput: "TwoZero")
+                .VerifyOperationTree(@"
+Test.Main()
```
I like the approach as well. Instead of having the operation tree inline though, can we instead have a partial type with constants like `TestBug11947_OperationTree` and the tests just get an extra line `.VerifyOperationTree(TestBug11947_OperationTree)` so that it doesn't clutter the binding tests?

AlekseyTs replied:
Yes, please do not merge multiple verifications in one statement.

---
In reply to: [65421246](https://github.com/dotnet/roslyn/pull/11680#discussion_r65421246) [](ancestors = 65421246)

mavasani replied:
I think we will likely have following category of tests:

1. Inline VerifyOperationTree tests for small method bodies - this will not clutter the existing binding tests as the tree will supposedly be pretty small.
2. VerifyOperationTree tests on large method bodies - I agree that we should put the expected operation tree for such tests into a separate file.
3. Hand crafted tests with SyntaxNode based operation tree verification - similar to the ones already written by John and proposed by @AlekseyTs is another comment. I'll add a separate test framework helper to verify expected Operation tree given a syntax node.

genlu commented on src/Test/Utilities/Shared/Compilation/CompilationExtensions.cs:
```diff
@@ -148,5 +150,102 @@ internal static void VerifyAssemblyAliases(this Compilation compilation, params
+            foreach (var declaration in declarations.Where(d => d.DeclaredSymbol != null).OrderBy(d => d.DeclaredSymbol.ToDisplayString()))
+            {
+                if (declaration.DeclaredNode == null ||
+                    declaration.DeclaredSymbol == null || 
```
The first two checks seems redundant: `DeclaredNode` will never be null (there's an assert in the constructor) and elements with `DeclaredSymbol == null` are already filtered out from the IEnumerable

mavasani replied:
Thanks, will clean it up.

genlu commented on src/Test/Utilities/Shared/Compilation/CompilationExtensions.cs:
```diff
@@ -148,5 +150,102 @@ internal static void VerifyAssemblyAliases(this Compilation compilation, params
+                case SymbolKind.Field:
+                case SymbolKind.Event:
+                case SymbolKind.Method:
+                case SymbolKind.NamedType:
```
Should `NamedType` be excluded?

mavasani replied:
This matches http://source.roslyn.io/Microsoft.CodeAnalysis/R/0e1dc66f2235f48c.html

genlu commented on src/Test/Utilities/Shared/Compilation/OperationTreeVerifier.cs:
```diff
@@ -0,0 +1,1065 @@
+            LogString(nameof(IForLoopStatement));
+            VisitLoopStatementHeader(operation);
+
+            VisitChildOperation(operation.Condition, "Condition");
```
It seems you have basically reimplemented the traversal logic of all IOperations, is it possible to get a similar text output just by calling `base.Visit()` instead? Preferably, `OperationWalker` should be the single source of truth for traversing operation tree.

mavasani replied:
For quite a few cases we need the override the visitors to dump header strings for more readable dump. For example over here, when dumping the tree for a For loop statement, we dump "Condition: " header when visiting the condition operation. We can probably avoid doing so for nodes whose child trees don't need headers, I'll look into cleaning it up.

genlu commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -40,7 +39,51 @@ static void Main()
+        IArgument (Matching Parameter: value (Type: object)) (OperationKind.Argument)
+          IConversionExpression (ConversionKind.Cast, Implicit) (OperationKind.ConversionExpression, Type: object)
+            ILocalReferenceExpression: rn (OperationKind.LocalReferenceExpression, Type: E?)",
+"Main");
```
Sorry I didn't realize the PR is just for commenting on the text dump, you can ignore my other comments for now :)
The dump looks fine (at least for this specific test, it's easy to follow). This approach is great. I assume this means there's no need to craft more test code since we can reuse existing ones which should cover all grounds?

mavasani replied:
We'll probably have a mix of these tests.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -40,7 +39,51 @@ static void Main()
+    Local_3: r (Type: E)
+    Local_4: en (Type: E?)
+    Local_5: xn (Type: int?)
+    Local_6: rn (Type: E?)
```
What happens if something changes and the baseline doesn't match? How easy would it be to figure out where the difference and how to adjust the baseline?

mavasani replied:
When the test fails, we dump (1) the Assert.Equal failure message which shows the line/char with the first mismatch AND (2) The entire actual operation tree dump, so the user can copy it to a separate buffer and use their preferred diffing tool to compare actual and expected output.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -424,7 +467,123 @@ static void Main()
 
-            CompileAndVerify(source: source, expectedOutput: output);
+            CompileAndVerify(source: source, expectedOutput: output)
+                .VerifyOperationTree(@"
```
We should be able to do verification off of a compilation instance. Calling CompileAndVerify is not an option for error scenarios, but we should test them too.

mavasani replied:
Yes, this is just a convenience static helper that grabs the _compilation field from the CompilationVerifier and invokes `compilation.VerifyOperationTree`.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:
```diff
@@ -424,7 +467,123 @@ static void Main()
+            IArgument (Matching Parameter: value (Type: char)) (OperationKind.Argument)
+              ILiteralExpression (Text: 'n') (OperationKind.LiteralExpression, Type: char, Constant: n)
+        IBranchStatement (BranchKind.Break) (OperationKind.BranchStatement)",
+"Main");
```
These format looks fine to capture the status quo, but I find hard to believe that anyone would actually be able to verify that all these output is what we would expect to get.

mavasani replied:
I agree - we should primarily use full operation tree verification only for small method bodies. I extended this test only to demonstrate the dump format.

AlekseyTs commented on src/Test/Utilities/Shared/Compilation/CompilationExtensions.cs:
```diff
@@ -148,5 +150,102 @@ internal static void VerifyAssemblyAliases(this Compilation compilation, params
+            model.ComputeDeclarationsInNode(root, getSymbol: true, builder: declarations, cancellationToken: CancellationToken.None);
+
+            var actualTextBuilder = new StringBuilder();
+            foreach (var declaration in declarations.Where(d => d.DeclaredSymbol != null).OrderBy(d => d.DeclaredSymbol.ToDisplayString()))
```
Please replace all "var"s with types, it is impossible to make sense what is going on in this function.

AlekseyTs commented:
I am not sure if the "dumper" dumps all the properties of the nodes, including all parent interfaces, etc.
For symbols, we usually use ToTestDisplayString() output as the baseline.
I think we would want to have more granularity, i.e. should be able to dump a tree corresponding to a particular syntax.

mavasani commented:
> I am not sure if the "dumper" dumps all the properties of the nodes, including all parent interfaces, etc.

Yes, it ensures it dumps all properties of parent interfaces as well.

> For symbols, we usually use ToTestDisplayString() output as the baseline.

I'll change the dumper to use `ToTestDisplayString` instead of `ToDisplayString`

> I think we would want to have more granularity, i.e. should be able to dump a tree corresponding to a particular syntax.

I'll add a separate helper for operation tree verification for a given syntax node.

mavasani commented:
**Update:** This PR is now ready for review. I have addressed the PR feedback and made the following changes:

1. Reverted the inline VerifyOperationTree changes in binding tests.
2. Add VerifyOperationTree overload to verify operation tree for given syntax node. I have also moved existing VB IOperation tests to use this method.
3. Added couple of sample tests to show how we can verify operation tree for entire method body.
4. Changed OperationTreeVerifier to use base methods when it doesn't need any custom dumping logic for its child nodes.

Also tagging @dotnet/roslyn-compiler for review.

mavasani commented on src/Compilers/VisualBasic/Test/Semantic/Diagnostics/IOperationTests.vb:
```diff
@@ -57,63 +58,36 @@ End Module
-            Dim right2 As IOperation = add2.RightOperand
-            Assert.Equal(right2.Kind, OperationKind.LocalReferenceExpression)
-            Assert.Equal(DirectCast(right2, ILocalReferenceExpression).Local.Name, "y")
+            comp.VerifyOperationTree(nodes(1), expectedOperationTree:="
```
Operation tree verification for given syntax node.

mavasani commented on src/Compilers/VisualBasic/Test/Semantic/Diagnostics/IOperationTests.vb:
```diff
@@ -147,41 +121,94 @@ End Module
+                         </compilation>
+
+            CompilationUtils.CreateCompilationWithMscorlibAndVBRuntime(source) _
+                .VerifyOperationTree("Foo", "
```
Operation tree verification for given symbol.

mavasani commented on src/Test/Utilities/Shared/Compilation/OperationTreeVerifier.cs:
```diff
@@ -0,0 +1,1046 @@
+
+            LogLocals(operation.Locals);
+
+            base.VisitBlockStatement(operation);
```
@genlu Moved to using base visitors where we don't do custom logging for child nodes.

mavasani commented:
Closing this PR in favor of https://github.com/dotnet/roslyn/pull/11918
